### PR TITLE
[v0.26.0][GPU] Integrate DeepSeek remote-experts refactor

### PR DESCRIPTION
**Pull request by jiaran-king**
## Purpose

This integration PR carries the exact GPU changes developed in #176 into
`update/v0.26.0`.

The goal is to let the GPU and NPU vLLM 0.26 upgrade work converge on the
integration branch before `update/v0.26.0` is submitted to `main` as one
combined change.

Source identity:

- Source PR: #176
- Exact head: `cb0a40a99a63c26de6edf8fd75f99dfa22d7b2d7`
- Target branch at submission:
  `update/v0.26.0` at `9d5bffc5c1fcca2157befbc8ca4820cb6990344a`
- No additional code changes or experiment artifacts are included.

## Issue

- Related PR: #176
- This PR does not close #176; it provides the GPU portion for the v0.26
  integration branch.

### In scope

- Upgrade the pinned GPU runtime contract from vLLM `0.19.1` to exact
  vLLM `0.26.0`.
- Use GPU ModelRunner V1 (`VLLM_USE_V2_MODEL_RUNNER=0`).
- Refactor DeepSeek MoE communication from the whole-MLP boundary to the
  remote-experts boundary.
- Preserve native `Model.forward`, `DecoderLayer.forward`, and
  `DeepseekV2MoE.forward`.
- Support both Attention-side and FFN-side gate placement.
- Keep routed/shared expert ownership on the FFN model runners.
- Reuse the existing connector data-path methods and ordered P2P custom ops.
- Align DBO and CUDA Graph execution with vLLM `0.26.0`.
- Retain compatibility adaptations in the existing NPU DeepSeek helper modules
  where required by the shared model contract.

### Out of scope

- The NPU/Ascend vLLM 0.26 runtime upgrade and hardware validation; these will
  be integrated separately on `update/v0.26.0`.
- ModelRunner V2.
- Quantized MoE and EPLB validation.
- Full 1,319-sample GSM8K evaluation.

## Implementation Notes

The GPU implementation keeps the native DeepSeek model hierarchy and introduces
a parameter-free Attention experts proxy at the remote-experts boundary.

For Attention-side gate placement:

```text
native DeepseekV2MoE.forward
→ native gate
→ parameter-free AFD experts proxy
→ existing send_attn_output(hidden states, optional router logits)
→ native FFN external-router MoE
→ existing send_ffn_output(final hidden states)
```

For FFN-side gate placement:

```text
native DeepseekV2MoE.forward
→ parameter-free internal-router proxy
→ existing send_attn_output(hidden states)
→ native FFN gate + experts
→ existing send_ffn_output(final hidden states)
```

No new experts-specific public connector API or second request/response protocol
is introduced.

The two changed files under
`afd_plugin/model_executor/models/npu/` adapt existing NPU helper code to the
shared DeepSeek model interface. They do not constitute the NPU vLLM 0.26
runtime upgrade or NPU hardware validation.

## Test Plan

- Run CPU-safe unit and package tests.
- Validate both gate-placement paths on GPU.
- Validate eager and CUDA Graph execution.
- Validate TP2/EP2 and DP2/EP2 paths.
- Validate representative 2-Attention/2-FFN DBO decode execution.
- Keep NPU upgrade and NPU hardware validation in the separate NPU integration
  work.

## Test Result

Current exact-head validation:

- Focused unit suite: 92 passed.
- Complete unit suite: 353 passed, 29 skipped.
- Ruff, formatting, and diff checks: passed.
- `1A1F` FFN-side gate eager: passed.
- `1A1F` Attention-side gate eager: passed.
- TP2/EP2 CUDA Graph: passed.
- Representative `2A2F` DBO + CUDA Graph completions: 8/8 matched.
- Exact-head decode-only profiling confirmed CUDA Graph replay and DBO
  two-ubatch behavior.

## Docs Impact

- Updated `docs/design/module/model_integration.md`.
- Updated the exact vLLM dependency/version contract and lockfile.

---

<details>
<summary>Essential PR Checklist</summary>

- [x] Purpose is clear and linked to #176.
- [x] Scope is bounded.
- [x] Compatibility with vLLM v0.19.1 is considered: this integration branch
      intentionally moves to exact vLLM v0.26.0 and does not claim dual-version
      support.
- [x] No changes are made to the vLLM source checkout.
- [x] Plugin-owned classes and explicit model/runner contracts are used.
- [x] Compatibility patches are isolated and covered by focused tests.
- [x] Imports remain CPU-safe; CUDA-heavy work is GPU-gated.
- [x] Available CPU and GPU validation evidence is documented.
- [x] Documentation impact is stated.

</details>

### CODE REVIEW

**Comment by jiaran-king** [comment]
### PR182 validation summary

Validated target: `update/v0.26.0@ff6ebd42ecc1e44cda11356577c2bc6cd8997bc7`  
Environment: NVIDIA L20X, DeepSeek-V2-Lite, vLLM `0.26.0`, ModelRunner V1

| Test area | Conditions covered | Result | Notes |
|---|---|---:|---|
| Installation and plugin loading | Locked non-editable installation; plugin loaded from the exact checkout | ✅ PASS | Commit/tree and runtime identity were verified |
| Non-GPU validation | Unit and runner tests | ✅ PASS | 363 passed; 29 expected NPU skips |
| Official GPU E2E | 20 collected GPU nodes | ✅ 20/20 PASS at executed scope | No failed, errored, or unexpectedly skipped JUnit node |
| Serving and API | Basic completions, usage statistics, invalid-model handling | ✅ PASS | Requests and expected error handling completed normally |
| 1A1F | Eager and `FULL_DECODE_ONLY` CUDA Graph | ✅ PASS | All requests completed; official eager/Graph comparison passed |
| DP2/EP2 | Eager/Graph × DBO OFF/ON | ✅ PASS | All configurations completed inference successfully |
| TP2/EP2 | Eager and CUDA Graph | ✅ PASS | Official TP tests passed; supplemental TP collective observability remains under review |
| DBO execution | Eager and Graph DBO-enabled cases | ✅ PASS | Evidence confirmed real ubatch stages 0 and 1 |
| CUDA Graph | 1A1F, DP2/EP2, and TP2/EP2 | ✅ PASS | Graph capture, launch, and structured replay were observed |
| Profiler | Eager and Graph; Attention and FFN roles | ✅ PASS | Expected profiler traces were generated and parsed |
| GSM8K eager | 50-item evaluation | ✅ PASS (50-item scope) | Strict/flexible exact match: `0.26`; effective threshold: `0.15` |
| GSM8K Graph | 50-item evaluation | ✅ PASS (50-item scope) | Strict exact match: `0.26`; flexible extract: `0.24` |
| Resource cleanup | Processes, ports, and selected GPUs | ✅ PASS | No run-owned process, port, or GPU leak remained |

### Follow-up observation

Supplemental cross-configuration comparisons observed differences in selected-token
logprobs and, for some requests, generated text/token IDs across eager/Graph,
DP/TP, and DBO OFF/ON configurations.

These observations did not cause request failures, missing responses, hangs, or
resource leaks. In the 50-item GSM8K evaluation, eager and Graph had the same
aggregate strict accuracy (`0.26`). The practical impact, root cause, and
PR182-specific attribution remain under investigation.

Post-response shutdown/teardown messages are tracked separately as known
non-blocking observations because all expected responses and JUnit outcomes had
already completed and cleanup succeeded.
